### ISSUE TITLE
qrcode: regression test No. 25 fails on Aarch64

### ISSUE THREAD

**Issue by tomoaki0705** [opencv/opencv#12926]
##### System information (version)

- OpenCV => 3.4 / master branch 
- Operating System / Platform => Jetson TX2 Aarch64 Ubuntu 16.04 64bit
- Compiler => GCC 5.4.0

##### Detailed description

I wrote PR #12905 , but even after mergin this PR, the test fails on Jetson TX2 (Aarch64)

```
$ ./bin/opencv_test_objdetect --gtest_filter=*QRC*/25
CTEST_FULL_OUTPUT
OpenCV version: 3.4.3-dev
OpenCV VCS version: 3.4.3-357-gd3f75df-dirty
Build type: Release
Compiler: /usr/bin/c++  (ver 5.4.0)
Parallel framework: pthreads
CPU features: NEON FP16
OpenCL is disabled
Note: Google Test filter = *QRC*/25
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Objdetect_QRCode
[ RUN      ] Objdetect_QRCode.regression/25, where GetParam() = "russian.jpg"
/home/nvidia/opencv-fork/modules/objdetect/test/test_qrcode.cpp:69: Failure
Value of: detectQRCode(src, corners)
  Actual: false
Expected: true
[  FAILED  ] Objdetect_QRCode.regression/25, where GetParam() = "russian.jpg" (120 ms)
[----------] 1 test from Objdetect_QRCode (120 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (120 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Objdetect_QRCode.regression/25, where GetParam() = "russian.jpg"

 1 FAILED TEST
```

#12905 was fixing the "decode" part, but this failure is on  "detect" part.
Looking back , the detection was failed on [this](https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L319) if section
https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L319
Tracing back, the different behaviour comes from [here](https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L202-L205)
https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L202-L205
This is computing the ratio of black and white, which should follow the QR code spec 1:1:3:1:1.
Unfortunately, this test case No. 25 is quite diagonal, so the ratio error becomes bit large, large enough above eps.

The difference of the binary image is very small.
Probably this difference comes from ```adaptiveThreshold```
https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L69
It's using ```ADAPTIVE_THRESH_GAUSSIAN_C```, so probably it has float rounding inside, and thus the binary image doesn't become "bit-exact"

I also loosen the ```eps_horizontal``` default_value from 0.1 to 0.11 and 0.2 [here](https://github.com/opencv/opencv/blob/ab2c16b2e55acaf193815e43475eb11670c45d60/modules/objdetect/src/qrcode.cpp#L28)
In either case, this test No. 25 passes, but that causes false positive on other test cases.

So as a result of looking in this implementation, this test No.25 is on the edge of this algorithm, very edge so that even a rounding error could cause a different behavior
Now, I'd like to discuss about how to work with this issue.
1. ignore and keep this test fail on Aarch64
2. disable/skip this test on platform other than x86 (or totally remove this test case)
3. fix it from the source (this at least includes the implementation of bit-exact ```adaptiveThreshold```)

##### Steps to reproduce

Run ```opencv_test_objdetect``` on Jetson TX2 (probably on any Aarch64 platform)

**Comment by allnes** [comment]
@tomoaki0705 please send binary image obtained after `adaptiveThreshold` in detection

**Comment by tomoaki0705** [comment]
Here you go.
![binarized_aarch64](https://user-images.githubusercontent.com/3737315/47505835-d9cd8e80-d8a9-11e8-8cf9-391563abab12.png)

Just to make sure, the md5sum has
```
$ md5sum binarized_aarch64.png
de4f799bb2b5bf8f3c208fe4afde4b12  binarized_aarch64.png
```

**Comment by allnes** [comment]
@tomoaki0705 Try to use data from this [https://github.com/opencv/opencv_extra/pull/537](https://github.com/opencv/opencv_extra/pull/537)

**Comment by tomoaki0705** [comment]
@allnes , thank you very much !
I confirmed.  It passed.

```
$ ./bin/opencv_test_objdetect --gtest_filter=*QRC*/25
CTEST_FULL_OUTPUT
OpenCV version: 3.4.3-dev
OpenCV VCS version: 3.4.3-412-gdefeda2-dirty
Build type: Release
Compiler: /usr/bin/c++  (ver 5.4.0)
Parallel framework: pthreads
CPU features: N/A
OpenCL is disabled
Note: Google Test filter = *QRC*/25
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Objdetect_QRCode
[ RUN      ] Objdetect_QRCode.regression/25, where GetParam() = "russian.jpg"
[       OK ] Objdetect_QRCode.regression/25 (444 ms)
[----------] 1 test from Objdetect_QRCode (444 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (444 ms total)
[  PASSED  ] 1 test.
```

Just for information
```
$ md5sum /home/nvidia/opencv_extra/testdata/cv/qrcode/russian.jpg
0364fcbb20d8e43350ab1951bc2ee145  /home/nvidia/opencv_extra/testdata/cv/qrcode/russian.jpg
```
On Jetson TX2

**Comment by tomoaki0705** [comment]
Additionally, I tried on Raspberry Pi 3, too.
Both current 3.4 branch and your update, work fine.  No regression. Good.

**Comment by tomoaki0705** [comment]
fixed by opencv/opencv_extra#537
